### ISSUE TITLE
feat: prompt lookup speculative decoding

### ISSUE THREAD

**Issue by ttupper92618** [Foxlight-Foundation/Skulk#142]
## Summary

Skulk generates one token per model forward pass. Prompt lookup speculative decoding generates multiple candidate (\"draft\") tokens by n-gram matching the current context against the input prompt — zero inference cost, no draft model — and verifies them in a single batched forward pass. When the draft is correct, multiple tokens are accepted from one model call.

In Skulk's distributed pipeline each forward pass incurs inter-node latency. Batching K draft tokens into one verification pass amortizes that cost; accepted tokens are essentially free.

Reference: [Look-Ahead Decoding / Prompt Lookup Decoding, 2023](https://arxiv.org/abs/2312.11462).

---

## Current state

- `generate.py:1062-1107` (`_stream_generate_without_lookahead`) generates exactly 1 token per iteration
- `MlxGenerationResponse` already has a `from_draft: bool = False` field (never set)
- No speculative mechanism of any kind exists today

---

## Key files

| File | Role |
|---|---|
| `src/exo/worker/engines/mlx/generator/generate.py:1011-1122` | Decode loop — where draft verification hooks in |
| `src/exo/worker/runner/llm_inference/batch_generator.py:73` | `_EngineTask.all_prompt_tokens` — full prompt token array, available throughout generation |
| `src/exo/shared/types/text_generation.py:91-122` | `TextGenerationTaskParams` — per-request config lives here |
| `src/exo/worker/engines/mlx/cache.py` | `KVPrefixCache`, `RotatingKVCache` — cache offset management |

---

### Phase 1 — N-gram lookup engine
**New file:** `src/exo/worker/engines/mlx/generator/prompt_lookup.py`

```python
def find_candidate_tokens(
    context_tokens: list[int],   # all tokens so far (prompt + generated)
    num_pred_tokens: int,        # K — how many draft tokens to return
    ngram_size: int = 2,         # trigger window length
) -> list[int] | None:
```

- Take the last `ngram_size` tokens as the query trigger
- Scan `context_tokens` from most-recent to oldest for the same n-gram
- Return up to `num_pred_tokens` tokens following the first match, or `None`
- Pure Python / numpy — runs on CPU, negligible overhead

### Phase 2 — Verification pass

Hook into `_stream_generate_without_lookahead` (and the `stream_generate` path) immediately after the sampler produces token `t+1`:

1. Call `find_candidate_tokens` for `K` draft candidates
2. If `None`, continue with the normal 1-token step
3. If candidates found, feed all `K` tokens to the model in a **single batched forward pass**: `model(drafts[None], cache=prompt_cache)` → logits shape `[1, K, vocab]`
4. Scan left-to-right: accept each position where `argmax(logits[i]) == drafts[i]`
5. Let `j` = first rejection index (or `K` if all accepted)
6. Yield tokens at positions `0..j` plus the model's own sample at position `j`
7. Net: `j+1` tokens from 1 forward pass, with `j ≥ 0`

### Phase 3 — KV cache handling

After the batched verification pass the cache contains state for all `K` positions, but only `j+1` were accepted.

- Save `cache.offset` before feeding draft tokens
- After determining acceptance count, set `cache.offset = saved_offset + j + 1`
- Verify `RotatingKVCache` and `ArraysCache` both support offset assignment (spike this first — it's the main unknown)
- If offset assignment is not clean, fall back to re-prefilling from the accepted prefix (slower but correct)

### Phase 4 — Configuration

Add to `TextGenerationTaskParams` (`text_generation.py:91`):

```python
prompt_lookup_num_tokens: int = 0        # 0 = disabled; typical: 5–10
prompt_lookup_ngram_size: int = 2        # trigger window; 2 is usually optimal
```

Wire through: `TextGeneration` task → `_start_next` / `_start_task` in `batch_generator.py` → generate loop.

### Phase 5 — API surface

- Accept `prompt_lookup_num_tokens` and `prompt_lookup_ngram_size` in chat completions request body as extension fields (same pattern as existing `top_k`, `seed`, `min_p`)
- Expose `x_skulk_draft_tokens_accepted` in the final chunk's usage block when enabled (for observability)
- Document in `website/docs/api-guide.md`

### Phase 6 — Tests

- Unit tests for `find_candidate_tokens`: exact match, partial match, no match, edge cases (context shorter than n-gram, match at boundary)
- Integration test: `prompt_lookup_num_tokens=0` produces identical token sequence as today's default
- Acceptance-rate smoke test: a prompt that repeats the same phrase should achieve `> 0` accepted draft tokens

---

## Design constraints

- **No regression on batch throughput.** If no n-gram match is found, the path must be identical to today (zero overhead). Gate the speculative step on a successful lookup before touching the model.
- **Greedy acceptance only.** Stochastic acceptance (the Leviathan et al. rejection-sampling scheme) is not needed for prompt lookup — we're targeting cheap structural wins, not correctness-preserving sampling under temperature.
- **Both generator paths.** `SequentialGenerator` and `BatchGenerator` both need support; `BatchGenerator` is the production priority.
- **Distributed compatibility.** The draft verification pass goes through the same pipeline as a normal forward pass — no special multi-node handling needed.

---

## Non-goals (separate issues)

- Draft-model speculative decoding
- MTP head injection (Qwen3-specific)
- Stochastic acceptance scheme

---

## Expected impact

| Workload | Expected gain |
|---|---|
| Structured output (JSON templates, tool calls) | 1.5–3× token throughput |
| Long doc Q&A / summarization | 1.3–2× |
| Short conversational turns | Neutral to modest |
| Distributed (multi-node): all of the above | Amplified — inter-node round-trip amortized |

---

🤖 Planned with [Claude Code](https://claude.com/claude-code)
